packages/core/app-extensions/src/actions/actionFactory.js

```diff
@@ -0,0 +1,84 @@
+import PropTypes from 'prop-types'
+import {useEffect, Suspense} from 'react'
+import {connect} from 'react-redux'
+import {LoadMask} from 'tocco-ui'
+import {consoleLogger} from 'tocco-util'
+
+import actionEmitter from '../actionEmitter'
+import DynamicAction from './components/DynamicAction'
+import {fetchActionPackages} from './modules/dynamicActions/actions'
+
+const renderLoader = () => <LoadMask />
+
+const dynamicActionFactory = actionPackage => {
+  if (!actionPackage) {
+    return null
+  }
+
+  const {packageName, appName} = actionPackage
+  return props => <DynamicAction {...props} packageName={packageName} appName={appName} />
+}
+
+const getActionPackage = (actionPackages, appId) => {
+  const actionPackage = actionPackages ? actionPackages.filter(a => a.actionName === appId) : null
+  return actionPackage && actionPackage.length > 0 ? actionPackage[0] : null
+}
+
+/**
+ * Create ActionComponent that instantiate a LazyAction dependend on the `appId`.
+ *
+ * @param {Object} actions
+ * ```
+ * const actions = {delete: React.lazy(() => import('./actions/Delete'))}
+ * ```
+ * @returns ActionComponent
+ */
+const actionFactory = actions => {
+  const LazyActionWrapper = props => {
+    const {appId, fetchActionPackages, actionPackages, ...actionProps} = props
+
+    useEffect(() => {
+      fetchActionPackages()
+    }, [fetchActionPackages])
+
+    // wait until all actions have beend loaded
+    if (typeof actionPackages === 'undefined' || actionPackages === null) {
```

Comment:
I learned that a check to `undefined` is more save when used with `typeof` operator. Since then I always does it like this.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined#typeof_operator_and_undefined